File: c/reporting/13_plugin.c
/* LumasReport example 13 -- Register a custom function + custom exporter
 * DIRECTLY (native function pointers) instead of loading an external plugin
 * DLL. The engine now exports rptRegisterFunctionA / rptRegisterExporter, so
 * no rpt_testplugin.dll (nor rptLoadPlugin) is needed -- everything lives in
 * this single process against the single LumasPdf.dll.
 *
 *   PlugDouble(x)  ->  x * 2     (accepts an int or a float)   [expr function]
 *   exporter target 100  ->  writes a marker string to a file  [custom target]
 *
 * The report calls {{expr: PlugDouble(21)}} (-> 42) and PlugDouble(2.5) (-> 5).
 * After exporting PDF/TXT we invoke the custom exporter via rptExport(job,100,..).
 */
#include <stdio.h>
#include <string.h>
#include "lumaspdf.h"

#define PDF_DEMO_KEY "LUMAS-LumasReportExamples-DD5D40E0"
#define RPT_DEMO_KEY "LRPT2-TFIwMgECAAIAAgIAAQADBAD___9_BAEAAAUBAAAGCAAAAAAAAAAAAAcQAEx1bWFzUnB0RXhhbXBsZXM.sneCg6LX_J59AcqzUZPx1WqQ7jGQZyji4z0WMm_vIoycNySw12vGVoKV76gk3-yQtx3LfZZepGxoDdQAgf5MAA"

/* ---- Custom expression function: PlugDouble(x) = x * 2 (stdcall ABI) ------- */
static int __stdcall PlugDouble(void* User, TRptCValue* Args, int NArgs, TRptCValue* ResultV) {
    (void)User;
    if (NArgs != 1) return -1;
    switch (Args[0].Kind) {
        case 2: /* vkInt   */ ResultV->Kind = 2; ResultV->I = Args[0].I * 2;     break;
        case 3: /* vkFloat */ ResultV->Kind = 3; ResultV->F = Args[0].F * 2.0;   break;
        default: return -2;
    }
    return 0;
}

/* ---- Custom exporter for target id 100 (stdcall ABI) ---------------------- */
static int __stdcall PlugExport(void* User, void* Job, const char* Path) {
    FILE* f;
    (void)User; (void)Job;
    if (!Path) return -1;
    f = fopen(Path, "wb");
    if (!f) return -1;
    fputs("PLUGIN:OK -- custom exporter via rptRegisterExporter (no external plugin DLL)", f);
    fclose(f);
    return 0;
}

static void WriteText(const char* path, const char* content) {
    FILE* f = fopen(path, "wb");
    if (f) { fputs(content, f); fclose(f); }
}

static void DumpRptError(TRPT eng) {
    TRptErrorInfoC info;
    memset(&info, 0, sizeof(info));
    if (rptGetLastError(eng, &info) != 0 && info.Code != 0)
        printf("  ! rpt error %d [%.16s] at %.64s: %.256s\n", info.Code, info.Module_, info.Location, info.Msg);
}

static int BootEngine(PPDF* pdf, TRPT* eng) {
    *pdf = pdfNewPDF();
    if (!*pdf) { printf("pdfNewPDF failed\n"); return 0; }
    pdfSetLicenseKey(*pdf, PDF_DEMO_KEY);
    rptSetRptLicenseKeyA(*pdf, RPT_DEMO_KEY);
    *eng = rptCreateEngineA(*pdf, NULL);
    if (!*eng) { printf("rptCreateEngine failed\n"); return 0; }
    return 1;
}

static const char* BuildXml(void) {
    return
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        "<report name=\"Plugin\" tagLangVersion=\"1\">\n"
        " <page width=\"210\" height=\"297\" marginLeft=\"15\" marginTop=\"15\" marginRight=\"15\" marginBottom=\"15\"/>\n"
        " <bands>\n"
        "  <band kind=\"reportheader\" name=\"rh\" height=\"24\">\n"
        "   <text name=\"p1\" x=\"0\" y=\"0\"  w=\"180\" h=\"8\" fontSize=\"16\">PlugDouble(21) = {{expr: PlugDouble(21) }}</text>\n"
        "   <text name=\"p2\" x=\"0\" y=\"10\" w=\"180\" h=\"8\" fontSize=\"12\">PlugDouble(2.5) = {{expr: PlugDouble(2.5) }}</text>\n"
        "  </band>\n"
        " </bands>\n"
        "</report>\n";
}

int main(void) {
    PPDF pdf; TRPT eng; TRPTJOB job;
    const char* lrpt = "13_plugin.lrpt";
    const char* outPdf = "13_plugin.pdf";
    const char* outTxt = "13_plugin.txt";
    const char* outCustom = "13_custom.out";
    if (!BootEngine(&pdf, &eng)) return 1;

    /* Register the custom function + exporter DIRECTLY -- BEFORE opening. */
    if (rptRegisterFunctionA(eng, "PlugDouble", 1, 1, (void*)&PlugDouble, NULL) == 0) {
        printf("rptRegisterFunction failed\n"); DumpRptError(eng); goto cleanup;
    }
    if (rptRegisterExporter(eng, 100, (void*)&PlugExport, NULL) == 0) {
        printf("rptRegisterExporter failed\n"); DumpRptError(eng); goto cleanup;
    }
    printf("registered function PlugDouble(x)=x*2 and exporter target 100 (no plugin DLL)\n");

    WriteText(lrpt, BuildXml());
    job = rptOpenReportA(eng, lrpt);
    if (!job) { printf("open failed\n"); DumpRptError(eng); goto cleanup; }
    if (rptRender(job) == 0) { printf("render failed\n"); DumpRptError(eng); goto closejob; }
    if (rptExportA(job, RPT_EXP_PDF, outPdf) == 0) { printf("export PDF failed\n"); DumpRptError(eng); goto closejob; }
    if (rptExportA(job, RPT_EXP_TEXT, outTxt) == 0) { printf("export TEXT failed\n"); DumpRptError(eng); goto closejob; }
    /* Invoke the custom exporter (target 100). */
    if (rptExportA(job, 100, outCustom) == 0) { printf("custom export failed\n"); DumpRptError(eng); goto closejob; }
    printf("wrote %s\n", outPdf);
    printf("wrote %s (via custom exporter)\n", outCustom);
    printf("OK\n");
closejob:
    rptCloseReport(job);
cleanup:
    rptDeleteEngine(eng);
    pdfDeletePDF(pdf);
    return 0;
}

File: csharp/reporting/13_plugin.cs
//  LumasReport example 13 -- Custom function + custom exporter, NO plugin DLL (C# port)
//  Instead of loading rpt_testplugin.dll via rptLoadPlugin, this registers the
//  very same PlugDouble(x)=x*2 expression function AND a custom export target
//  directly on the engine through native-callback delegates:
//     rptRegisterFunction(eng, "PlugDouble", 1, 1, fnPtr, 0)   // {{expr: PlugDouble(21)}}
//     rptRegisterExporter (eng, 100, expPtr, 0)                // rptExport(job, 100, path)
//  Both callbacks use the engine's stdcall C ABI. The delegates are kept alive in
//  static fields so the GC cannot collect the thunks. No external plugin DLL is used.
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class Ex13
{
    const string PDF_DEMO_KEY = "LUMAS-LumasReportExamples-DD5D40E0";
    const string RPT_DEMO_KEY = "LRPT2-TFIwMgECAAIAAgIAAQADBAD___9_BAEAAAUBAAAGCAAAAAAAAAAAAAcQAEx1bWFzUnB0RXhhbXBsZXM.sneCg6LX_J59AcqzUZPx1WqQ7jGQZyji4z0WMm_vIoycNySw12vGVoKV76gk3-yQtx3LfZZepGxoDdQAgf5MAA";

    // TRptCValue ordinals and field offsets: Kind@0, B@4, I@8, F@16, S@24.
    const int VK_INT = 2;
    const int VK_FLOAT = 3;
    const int OFF_I = 8;
    const int OFF_F = 16;
    // Custom export target id (must be >= 100).
    const int TARGET_CUSTOM = 100;
    const string EXPORT_MARKER =
        "PLUGIN:OK -- custom exporter via rptRegisterExporter (no external plugin DLL)";

    // stdcall callback ABIs. Args/ResultV are PRptCValue; Job/Path are opaque + PAnsiChar.
    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
    delegate int TRptUserFn(IntPtr User, IntPtr Args, int NArgs, IntPtr ResultV);
    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
    delegate int TRptExporterFn(IntPtr User, IntPtr Job, IntPtr Path);

    // Keep the delegates alive so the GC does not collect the marshalled thunks.
    static TRptUserFn _plugDoubleCb = PlugDouble;
    static TRptExporterFn _plugExportCb = PlugExport;

    // PlugDouble(x) -> x*2, preserving the numeric kind (int stays int, float stays float).
    static int PlugDouble(IntPtr user, IntPtr args, int nArgs, IntPtr resultV)
    {
        if (nArgs != 1 || args == IntPtr.Zero || resultV == IntPtr.Zero) return -1;
        int kind = Marshal.ReadInt32(args, 0);
        if (kind == VK_INT)
        {
            long v = Marshal.ReadInt64(args, OFF_I);
            Marshal.WriteInt32(resultV, 0, VK_INT);
            Marshal.WriteInt64(resultV, OFF_I, v * 2);
        }
        else if (kind == VK_FLOAT)
        {
            double f = BitConverter.Int64BitsToDouble(Marshal.ReadInt64(args, OFF_F));
            Marshal.WriteInt32(resultV, 0, VK_FLOAT);
            Marshal.WriteInt64(resultV, OFF_F, BitConverter.DoubleToInt64Bits(f * 2.0));
        }
        else return -2;
        return 0;
    }

    // Custom exporter for target id 100: write a marker file to the requested path.
    static int PlugExport(IntPtr user, IntPtr job, IntPtr path)
    {
        string p = Marshal.PtrToStringAnsi(path);
        if (string.IsNullOrEmpty(p)) return -1;
        try { File.WriteAllText(p, EXPORT_MARKER); } catch { return -2; }
        return 0;
    }

    static string BuildXml()
    {
        string s = "";
        s += "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
        s += "<report name=\"Plugin\" tagLangVersion=\"1\">\n";
        s += " <page width=\"210\" height=\"297\" marginLeft=\"15\" marginTop=\"15\" marginRight=\"15\" marginBottom=\"15\"/>\n";
        s += " <bands>\n";
        s += "  <band kind=\"reportheader\" name=\"rh\" height=\"24\">\n";
        s += "   <text name=\"p1\" x=\"0\" y=\"0\"  w=\"180\" h=\"8\" fontSize=\"16\">PlugDouble(21) = {{expr: PlugDouble(21) }}</text>\n";
        s += "   <text name=\"p2\" x=\"0\" y=\"10\" w=\"180\" h=\"8\" fontSize=\"12\">PlugDouble(2.5) = {{expr: PlugDouble(2.5) }}</text>\n";
        s += "  </band>\n";
        s += " </bands>\n";
        s += "</report>\n";
        return s;
    }

    static void Main()
    {
        IntPtr pdf, eng;
        if (!BootEngine(out pdf, out eng)) return;

        // Register BEFORE opening the report so the compiler resolves PlugDouble and
        // knows about the custom export target.
        IntPtr fnPtr = Marshal.GetFunctionPointerForDelegate(_plugDoubleCb);
        if (!LumasPdf.rptRegisterFunction(eng, "PlugDouble", 1, 1, fnPtr, IntPtr.Zero))
        {
            Console.WriteLine("rptRegisterFunction failed"); DumpRptError(eng); goto Cleanup;
        }
        IntPtr expPtr = Marshal.GetFunctionPointerForDelegate(_plugExportCb);
        if (!LumasPdf.rptRegisterExporter(eng, TARGET_CUSTOM, expPtr, IntPtr.Zero))
        {
            Console.WriteLine("rptRegisterExporter failed"); DumpRptError(eng); goto Cleanup;
        }
        Console.WriteLine("registered custom function PlugDouble/1 (x -> x*2)");
        Console.WriteLine("registered custom export target " + TARGET_CUSTOM + " (no external plugin DLL)");

        string dir = AppDomain.CurrentDomain.BaseDirectory;
        string lrpt = Path.Combine(dir, "13_plugin.lrpt");
        string outPdf = Path.Combine(dir, "13_plugin.pdf");
        string outTxt = Path.Combine(dir, "13_plugin.txt");
        string outCustom = Path.Combine(dir, "13_custom.out");
        File.WriteAllText(lrpt, BuildXml());

        IntPtr job = LumasPdf.rptOpenReportA(eng, lrpt);
        if (job == IntPtr.Zero) { Console.WriteLine("open failed"); DumpRptError(eng); goto Cleanup; }
        if (!LumasPdf.rptRender(job)) { Console.WriteLine("render failed"); DumpRptError(eng); goto CloseJob; }
        if (!LumasPdf.rptExportA(job, LumasPdfConsts.RPT_EXP_PDF, outPdf)) { Console.WriteLine("export PDF failed"); DumpRptError(eng); goto CloseJob; }
        if (!LumasPdf.rptExportA(job, LumasPdfConsts.RPT_EXP_TEXT, outTxt)) { Console.WriteLine("export TEXT failed"); DumpRptError(eng); goto CloseJob; }
        if (!LumasPdf.rptExportA(job, TARGET_CUSTOM, outCustom)) { Console.WriteLine("export CUSTOM failed"); DumpRptError(eng); goto CloseJob; }
        Console.WriteLine("wrote " + outPdf);
        Console.WriteLine("wrote " + outCustom + " (via custom exporter)");
        Console.WriteLine("OK");
    CloseJob:
        LumasPdf.rptCloseReport(job);
    Cleanup:
        LumasPdf.rptDeleteEngine(eng);
        LumasPdf.pdfDeletePDF(pdf);
    }

    static bool BootEngine(out IntPtr pdf, out IntPtr eng)
    {
        eng = IntPtr.Zero;
        pdf = LumasPdf.pdfNewPDF();
        if (pdf == IntPtr.Zero) { Console.WriteLine("pdfNewPDF failed"); return false; }
        LumasPdf.pdfSetLicenseKey(pdf, PDF_DEMO_KEY);
        LumasPdf.rptSetRptLicenseKeyA(pdf, RPT_DEMO_KEY);
        eng = LumasPdf.rptCreateEngineA(pdf, null);
        if (eng == IntPtr.Zero) { Console.WriteLine("rptCreateEngine failed"); return false; }
        return true;
    }

    static void DumpRptError(IntPtr eng)
    {
        IntPtr p = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(TRptErrorInfoC)));
        try
        {
            if (LumasPdf.rptGetLastError(eng, p))
            {
                var info = (TRptErrorInfoC)Marshal.PtrToStructure(p, typeof(TRptErrorInfoC));
                if (info.Code != 0)
                    Console.WriteLine("  ! rpt error " + info.Code + " [" + info.Module_ + "] at " + info.Location + ": " + info.Msg);
            }
        }
        finally { Marshal.FreeHGlobal(p); }
    }
}
